### DISCUSSION TITLE
Full ASCII transliteration in filenames

### DISCUSSION THREAD

**Discussion by keeganjk** [beetbox/beets#3968] [Q&A] [answered]
Is there a way that all file and directory names could use ASCII characters only, similarly to how there is a "sort name" for artists also using transliterated forms of album and track names, using `"` or `,` instead of "fancy" variants, replacing ligatures like `æ` with separated variants (`ae`) or accented characters with their ASCII equivalents (e.g. `ä` to `ae` or `é` to `e`), etc.? I'm sure MusicBrainz provides many of these (transliterations at least) and it should not be difficult to compute the common substitutions. If some of these are not possible, perhaps there are 3rd-party sources which `beets` could connect to for the ASCII variant.

**Comment by sampsyo** [comment, marked as answer]
It seems like [the `asciify_paths` config option](https://beets.readthedocs.io/en/stable/reference/config.html#asciify-paths) may be a good starting point, at least?

**Reply by keeganjk** [reply]
Yes, that seems like it will do the trick for this purpose.

This is unrelated, but how could I convert that similarly to lowercase? I tried looking for information about it on that page and I only saw a thing about sorting case-sensitive vs. case-insensitive, which I do not think is correct for converting all to a single case.

**Reply by sampsyo** [reply]
There's not a global option, but maybe give the [template functions](https://beets.readthedocs.io/en/stable/reference/pathformat.html#template-functions) a shot—there is a `%lower` function built in, for example.

**Reply by keeganjk** [reply]
Great, thank you—`%lower{}` works. 

Additionally: 

I want to essentially replace all special characters with `_` and then replace leftover spaces with `.`. However, what I'm trying results in replacement by underscores (one variation resulted in periods but no underscores). These make me believe that the replacements are not done sequentially, but it's like they're done globally so earlier steps will override later steps that involve the same characters. I have the follwing for `replace:`:

```
replace:
    '[\\/]': _
    '\.': _
    '[\x00-\x1f]': _
    '[\~@\#\$%\^!<>:\`"\;[]\?{}\*\|]': _
    "'": _
    '\s+$': ''
    '^\s+': ''
    '\s': .
```

I get `artist1_feat__artist2` when I'm looking for `artist1.feat_.artist2`. What should I do to achieve this?



Related to the original question: Is it possible to simply whitelist certain characters I want to allow rather than specifically replace a large portion of them?

**Reply by keeganjk** [reply]
My ultimate goal is just to make the filenames conform to the [POSIX Portable Filename Character Set](https://pubs.opengroup.org/onlinepubs/007904975/basedefs/xbd_chap03.html#tag_03_276). 

Perhaps the method of doing this is:
1. Replace all characters with ASCII equivalents
2. Replace all non-space characters with `_`
3. Replace all spaces with a remaining permitted delimiter (`.` or `-`, the latter of which may make more sense because `.` exists as `.flac`, etc.)

I may also wish to somehow make the sorting of multiple artists neater (the above steps will replace both `,` and `;` with `_` which may cause confusion), but the above steps are what is sought out in general.

Perhaps it would make sense to make a function to do the thing with multiple artists, but I do not know if this is possible from the YAML specifications alone.

**Reply by wisp3rwind** [reply]
On top of template functions, you might want to have a look at the [inline plugin](https://beets.readthedocs.io/en/stable/plugins/inline.html).
